### PR TITLE
Fix Python parser chunked handling with multiple Transfer-Encoding values

### LINKED ISSUE #4436: Differences in transfer encoding handling between the C and Python HTTP parsers
## Long story short
The C and Python parsers in aiohttp do not handle transfer encoding the same way, and neither behaves correctly. The C parser simply discards any value that isn't `chunked` and therefore doesn't handle things like `gzip, chunked` (see #4435). The Python parser accepts any value that *contains* `chunked`, including invalid values like `notchunked`, and parses (or attempts to parse) the payload as a chunked response.

## Expected behaviour
The C and Python implementations behave the same and according to the HTTP specification, i.e. presumably throw an error on TE values they do not support or that are invalid (though the standard does not explicitly mention that as far as I can see).

## Actual behaviour
The two implementations differ and do not conform to the standard.

## Steps to reproduce
A simple server and client illustrating this issue can be found in [this gist](https://gist.github.com/JustAnotherArchivist/ba23cb434810254a7d55ec893e45e1d4). I wrote a single test client and server for both this issue and #4435; only the first line is relevant to this issue, and the others are not included in the output below.

Behaviour of the C parser:
```
> python3 client.py | head -1
b'1\r\nT\r\n3\r\nest\r\n0\r\n\r\n'
```

Behaviour of the Python parser:
```
> AIOHTTP_NO_EXTENSIONS=1 python3 client.py | head -1
b'Test'
```

## Your environment
I tested this with aiohttp 2.3.10 and Python 3.6.9 on Debian, but based on the current aiohttp code, the behaviour should still be the same on the current versions.

### PR DESCRIPTION
**Pull request by Dreamsorcerer**
Fixes #4436

### CODE REVIEW

**Review by bdraco — APPROVED** [review]

**Comment by bdraco** [comment]
LGTM sans the missing changelog message

File: aiohttp/http_parser.py [hunk 1]
```diff
@@ -723,6 +728,9 @@ def parse_message(self, lines: List[bytes]) -> RawResponseMessage:
             chunked,
         )
 
+    def _is_chunked_te(self, te: str) -> bool:
+        return te.rsplit(",", maxsplit=1)[-1].strip(" \t").lower() == "chunked"
```
Comment (Dreamsorcerer) [inline aiohttp/http_parser.py:734]:
```suggestion
        # https://www.rfc-editor.org/rfc/rfc9112#section-6.3-2.4.2
        return te.rsplit(",", maxsplit=1)[-1].strip(" \t").lower() == "chunked"
```

File: tests/test_http_parser.py [hunk 2]
```diff
@@ -1191,6 +1207,25 @@ async def test_http_response_parser_bad_chunked_strict_c(
...
 
 
+async def test_http_response_parser_notchunked(
+    response: HttpResponseParser,
+) -> None:
+    text = b"HTTP/1.1 200 OK\r\nTransfer-Encoding: notchunked\r\n\r\n1\r\nT\r\n3\r\nest\r\n0\r\n\r\n"
+    messages, upgrade, tail = response.feed_data(text)
+    response.feed_eof()
+
+    assert await messages[0][1].read() == b"1\r\nT\r\n3\r\nest\r\n0\r\n\r\n"
```
Comment (Dreamsorcerer) [inline tests/test_http_parser.py:1220]:
```suggestion
    https://www.rfc-editor.org/rfc/rfc9112#section-6.3-2.4.2
    assert await messages[0][1].read() == b"1\r\nT\r\n3\r\nest\r\n0\r\n\r\n"
```

File: tests/test_http_parser.py [hunk 3]
```diff
@@ -1191,6 +1207,25 @@ async def test_http_response_parser_bad_chunked_strict_c(
...
+
+
+async def test_http_response_parser_last_chunked(
+    response: HttpResponseParser,
+) -> None:
+    text = b"HTTP/1.1 200 OK\r\nTransfer-Encoding: not, chunked\r\n\r\n1\r\nT\r\n3\r\nest\r\n0\r\n\r\n"
+    messages, upgrade, tail = response.feed_data(text)
+
+    assert await messages[0][1].read() == b"Test"
```
Comment (Dreamsorcerer) [inline tests/test_http_parser.py:1230]:
```suggestion
    # https://www.rfc-editor.org/rfc/rfc9112#section-6.3-2.4.2
    assert await messages[0][1].read() == b"Test"
```

File: tests/test_http_parser.py [hunk 4]
```diff
@@ -532,6 +533,21 @@ def test_request_te_chunked123(parser: HttpRequestParser) -> None:
...
 
+async def test_request_te_last_chunked(parser: HttpRequestParser) -> None:
+    text = b"GET /test HTTP/1.1\r\nTransfer-Encoding: not, chunked\r\n\r\n1\r\nT\r\n3\r\nest\r\n0\r\n\r\n"
+    messages, upgrade, tail = parser.feed_data(text)
+    assert await messages[0][1].read() == b"Test"
+
+
+def test_request_te_first_chunked(parser: HttpRequestParser) -> None:
+    text = b"GET /test HTTP/1.1\r\nTransfer-Encoding: chunked, not\r\n\r\n1\r\nT\r\n3\r\nest\r\n0\r\n\r\n"
+    with pytest.raises(
```
Comment (Dreamsorcerer) [inline tests/test_http_parser.py:546]:
```suggestion
    # https://www.rfc-editor.org/rfc/rfc9112#section-6.3-2.4.3
    with pytest.raises(
```

File: tests/test_http_parser.py [hunk 5]
```diff
@@ -532,6 +533,21 @@ def test_request_te_chunked123(parser: HttpRequestParser) -> None:
         parser.feed_data(text)
 
 
+async def test_request_te_last_chunked(parser: HttpRequestParser) -> None:
+    text = b"GET /test HTTP/1.1\r\nTransfer-Encoding: not, chunked\r\n\r\n1\r\nT\r\n3\r\nest\r\n0\r\n\r\n"
+    messages, upgrade, tail = parser.feed_data(text)
+    assert await messages[0][1].read() == b"Test"
```
Comment (Dreamsorcerer) [inline tests/test_http_parser.py:540]:
```suggestion
    # https://www.rfc-editor.org/rfc/rfc9112#section-6.3-2.4.3
    assert await messages[0][1].read() == b"Test"
```

File: aiohttp/http_parser.py [hunk 6]
```diff
@@ -638,6 +638,11 @@ def parse_message(self, lines: List[bytes]) -> RawRequestMessage:
             url,
         )
 
+    def _is_chunked_te(self, te: str) -> bool:
+        if te.rsplit(",", maxsplit=1)[-1].strip(" \t").lower() == "chunked":
+            return True
+        raise BadHttpMessage("Request has invalid `Transfer-Encoding`")
```
Comment (Dreamsorcerer) [inline aiohttp/http_parser.py:645]:
```suggestion
        # https://www.rfc-editor.org/rfc/rfc9112#section-6.3-2.4.3
        raise BadHttpMessage("Request has invalid `Transfer-Encoding`")
```

File: tests/test_http_parser.py [hunk 7]
```diff
@@ -1191,6 +1209,27 @@ async def test_http_response_parser_bad_chunked_strict_c(
...
 
 
+async def test_http_response_parser_notchunked(
+    response: HttpResponseParser,
+) -> None:
+    text = b"HTTP/1.1 200 OK\r\nTransfer-Encoding: notchunked\r\n\r\n1\r\nT\r\n3\r\nest\r\n0\r\n\r\n"
+    messages, upgrade, tail = response.feed_data(text)
+    response.feed_eof()
+
+    https://www.rfc-editor.org/rfc/rfc9112#section-6.3-2.4.2
```
Comment (Dreamsorcerer) [inline tests/test_http_parser.py]:
```suggestion
    # https://www.rfc-editor.org/rfc/rfc9112#section-6.3-2.4.2
```

**Pull request merged**